Regression in test suite of probor-0.2.1 in Rust 1.15

https://github.com/tailhook/probor at d0be60ec37ded2eee0f851887e2d7c06fe2e7021

```
101 brian@ip-10-145-43-250:~/dev/probor/rust⟫ rustc +beta -Vv
rustc 1.15.0-beta.1 (d9a0f0df7 2016-12-19)
binary: rustc
commit-hash: d9a0f0df7051c603011d6b60fbdd155318fc47f3
commit-date: 2016-12-19
host: x86_64-unknown-linux-gnu
release: 1.15.0-beta.1
LLVM version: 3.9
```

```
brian@ip-10-145-43-250:~/dev/probor/rust⟫ cargo +beta test
   Compiling byteorder v1.0.0
   Compiling libc v0.2.18
   Compiling byteorder v0.5.3
   Compiling cbor-codec v0.5.0
   Compiling probor v0.2.1 (file:///mnt2/dev/probor/rust)
error[E0428]: a macro named `_probor_pattern` has already been defined in this module

error: aborting due to previous error

Build failed, waiting for other jobs to finish...
error: Could not compile `probor`.

```

cc @jseyfried I haven't seen one quite like this.

cc @tailhook

--- a/src/librustc_resolve/resolve_imports.rs
+++ b/src/librustc_resolve/resolve_imports.rs
@@ -21,6 +21,7 @@ use rustc::ty;
 use rustc::lint::builtin::PRIVATE_IN_PUBLIC;
 use rustc::hir::def_id::DefId;
 use rustc::hir::def::*;
+use rustc::util::nodemap::FxHashSet;
 
 use syntax::ast::{Ident, NodeId};
 use syntax::ext::base::Determinacy::{self, Determined, Undetermined};
@@ -728,7 +729,12 @@ impl<'a, 'b:'a> ImportResolver<'a, 'b> {
 
         let mut reexports = Vec::new();
         if module as *const _ == self.graph_root as *const _ {
-            reexports = mem::replace(&mut self.macro_exports, Vec::new());
+            let mut exported_macro_names = FxHashSet();
+            for export in mem::replace(&mut self.macro_exports, Vec::new()).into_iter().rev() {
+                if exported_macro_names.insert(export.name) {
+                    reexports.push(export);
+                }
+            }
         }
 
         for (&(ident, ns), resolution) in module.resolutions.borrow().iter() {

--- a/src/test/run-pass/auxiliary/issue_38715.rs
+++ b/src/test/run-pass/auxiliary/issue_38715.rs
@@ -1,0 +1,15 @@
+// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#[macro_export]
+macro_rules! foo { ($i:ident) => {} }
+
+#[macro_export]
+macro_rules! foo { () => {} }

--- a/src/test/run-pass/issue-38715.rs
+++ b/src/test/run-pass/issue-38715.rs
@@ -1,0 +1,20 @@
+// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// aux-build:issue_38715.rs
+
+// Test that `#[macro_export] macro_rules!` shadow earlier `#[macro_export] macro_rules!`
+
+#[macro_use]
+extern crate issue_38715;
+
+fn main() {
+    foo!();
+}

